### PR TITLE
Full replacement of paste.ee

### LINKED ISSUE #44: Switch away from paste.ee
### Role

We don't want to rely on MultiMC's related services, nor their API keys.

### Suggestion

Switch to a service such as paste.swurl.xyz, a tiny pastebin I made, because it's easier to use and doesn't require an API key.

### Benefit

MultiMC could revoke the API key to spite us.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### PR DESCRIPTION
**Pull request by ghost**
Closes: #44 

did some stuff will do more in a bit

todo:
- [x] Set up global app settings
- [x] Proper paste upload thing
- [ ] more pastebins
- [x] Make the Combo box set to the available pastebins (easy)

### CODE REVIEW

**Comment by ghost** [comment]
questions:
- how do you create global variables (i.e. QList) with this codebase? It sucks lol

**Comment by LennyMcLennington** [comment]
Do we need to add validation to the user input to make sure it's a valid URL or should we just leave it up to the user to make sure? Nothing bad will happen if they input something wrong other than them getting an error message when they try to upload. Probably would help prevent some errors related to the uri format that would be difficult to understand like one I got while testing it that was like `unknown protocol: ""` which was when I didn't have https:// at the beginning.
Also, what is meant by "Set up global app settings and list"?

**Comment by ghost** [comment]
By that, I meant getting the global default list initialized, and the whole application settings thing that this uses

Additionally, we can at least validate to ensure it's a valid URI. It'd be hard to validate its API.

File: launcher/net/PasteUpload.cpp [hunk 1]
```diff
@@ -8,44 +8,34 @@
...
 void PasteUpload::executeTask()
 {
-    QNetworkRequest request(QUrl("https://api.paste.ee/v1/pastes"));
+    QNetworkRequest request{QUrl(m_uploadUrl)};
     request.setHeader(QNetworkRequest::UserAgentHeader, BuildConfig.USER_AGENT_UNCACHED);
 
-    request.setRawHeader("Content-Type", "application/json");
-    request.setRawHeader("Content-Length", QByteArray::number(m_jsonContent.size()));
-    request.setRawHeader("X-Auth-Token", m_key.toStdString().c_str());
+    QHttpMultiPart *multiPart = new QHttpMultiPart{QHttpMultiPart::FormDataType};
```
Comment (ghost) [inline launcher/net/PasteUpload.cpp:24]: pacebin literally just uses the POST body itself, will need to test if it works properly here (not familiar with multipart so maybe it does)

**Comment by ghost** [comment]
I also lied it'd be easy. But you'd have to POST an example file, and it would probably waste some server space. Plus, all pastebins have different deletion APIs.

File: launcher/net/PasteUpload.cpp (hunk 1 above)
Comment (LennyMcLennington) [reply launcher/net/PasteUpload.cpp:24]: > pacebin literally just uses the POST body itself, will need to test if it works properly here (not familiar with multipart so maybe it does)

If you really want to be able to support both 0x0 and pacebin then you'll either have to change pacebin to support posting the file to the file parameter, or make it configurable.

**Comment by LennyMcLennington** [comment]
Yeah we don't need to check if it's a valid API but I think basic regex for a valid URL would be good.

I still don't know what you mean by "global default list". Can you explain why that would be necessary? The configuration option is already there and it just uses a URL that's provided by the user.

**Comment by ghost** [comment]
I was thinking we have a `QList` containing all the default options in the QComboBox, i.e. 0x0.st, paste.polymc.org... the latter for people who trust PolyMC services more than public ones. And so that I have peace of mind. I would *preferrably* like this to be configurable; however, thinking about it now, the UI can just be edited.

Also for regex:

`http(s?):\/\/(.*\.?).+\..+`

Seems to work.

File: launcher/net/PasteUpload.cpp (hunk 1 above)
Comment (ghost) [reply launcher/net/PasteUpload.cpp:24]: I will probably change pacebin, with a configurable `-DENABLE_MULTIPART` compile option. Should be easy to add, just use `mg_http_next_multipart()`.

**Comment by LennyMcLennington** [comment]
> Also for regex:
> 
> `http(s?):\/\/(.*\.?).+\..+`
> 
> Seems to work.

I don't think a hostname needs to have a dot in it, wouldn't `https?:\/\/.+` be ok?

**Comment by ghost** [comment]
I suppose it could, for localhost or we

**Comment by ghost** [comment]
pacebin has been updated to support multipart, and paste.polymc.org is now running pacebin.

**Comment by ghost** [comment]
This is ready now

File: launcher/net/PasteUpload.cpp [hunk 2]
```diff
@@ -59,7 +59,7 @@ void PasteUpload::downloadFinished()
         m_reply.reset();
         return;
     }
-    else if (statusCode != 200)
+    else if (statusCode != 200 && statusCode != 201)
```
Comment (LennyMcLennington) [inline launcher/net/PasteUpload.cpp:62]: What is the purpose of this

Comment (ghost) [reply launcher/net/PasteUpload.cpp:62]: pacebin (and maybe some other pastebins) return 201 Created upon successful paste creation.

**Comment by LennyMcLennington** [comment]
I'm gonna add a regex check to the input before merging this like we were discussing

**Comment by LennyMcLennington** [comment]
I feel like the way Qt validator thing works isn't that user friendly because the user just physically can't enter anything that doesn't fit the regex, so I am gonna put a note above the text entry explaining that functionality so people don't get confused.

**Pull request merged**